### PR TITLE
Massive review and cleanup of JCache

### PR DESCRIPTION
**Pull request by mbabker**
Unfortunately this PR does a lot.  It started just trying to clean up some of the tests but then several bits of dead code and bug fixes were needed to, so...
#### Summary of Changes

This PR makes the following changes:
- General changes
  - Creates an abstract `TestCaseCache` for all `JCacheStorage` handler tests and rewrites the test classes for those objects to use this test case (results in more consistent testing of all handlers when the environment supports it)
  - Standardizes doc blocks on all `JCache` classes for full consistency and removes some outdated or unneeded comments
  - Uses late static bindings where practical (enables better class inheritance)
- Bug fixes
  - `JCacheStorageMemcache::store()` was manipulating the handler's lifetime setting after being instantiated
  - `JCacheStorageRedis::store()` was manipulating the handler's lifetime setting after being instantiated, had some dead code (copy/paste from Memcache(d) not needed here), and used a hardcoded lifetime instead of the handler's configured lifetime

### CODE REVIEW

**Comment by brianteeman** [comment]
(a bit above my skill set but sounds like this might address some existing issues here https://issues.joomla.org/tracker/joomla-cms/?state=open&sort=issue&direction=desc&category=cache&stools-active=1 )<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9622">issues.joomla.org/joomla-cms/9622</a>.</sub>

**Comment by mbabker** [comment]
#9382 is included in here.  #8707 I've added the fix for (when `JCache` creates a handler the lifetime is already injected in as a constructor parameter and `JCacheStorage` handles conversion of minutes to seconds, as noted there this line just screws things up).  Ironically #6796 also makes that same change based on #2539.

**Comment by brianteeman** [comment]
Thanks for checking can you update those issues as needed

On 27 March 2016 at 19:12, Michael Babker <notifications@github.com> wrote:

> #9382 <https://github.com/joomla/joomla-cms/pull/9382> is included in
> here. #8707 <https://github.com/joomla/joomla-cms/issues/8707> I've added
> the fix for (when JCache creates a handler the lifetime is already
> injected in as a constructor parameter and JCacheStorage handles
> conversion of minutes to seconds, as noted there this line just screws
> things up). Ironically #6796
> <https://github.com/joomla/joomla-cms/pull/6796> also makes that same
> change based on #2539 <https://github.com/joomla/joomla-cms/issues/2539>.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/9622#issuecomment-202116504>
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

**Comment by brianteeman** [comment]
@mbabker I have updated the other issues. Thanks for reviewing them

**Comment by andrepereiradasilva** [comment]
I have tested this item :white_check_mark: successfully on 9228a8674c558e12d6af931e2b8785effa1bbbb8<br /><br />Wow that's really a massive review!

Did a code review (except the unit tests) and all seems fine.

Installed the patch turned on File cache and all seems fine.

Didn't test all the other cache handlers, but by looking at code changes, it seems fine and also corrects some bugs.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9622">issues.joomla.org/joomla-cms/9622</a>.</sub>

**Comment by alikon** [comment]
i've tested successfully using only the redis handler

**Comment by Kubik-Rubik** [comment]
@alikon Please use the Joomla! issue tracker to enter your test result. Thanks!

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on 9228a8674c558e12d6af931e2b8785effa1bbbb8<br /><br />Tested on Redis cache handler<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9622">issues.joomla.org/joomla-cms/9622</a>.</sub>

**Comment by brianteeman** [comment]
RTC thanks<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9622">issues.joomla.org/joomla-cms/9622</a>.</sub>

File: libraries/joomla/cache/storage/memcache.php [hunk 1]
```diff
@@ -52,92 +52,72 @@ public function __construct($options = array())
...
+		$this->_compress   = $config->get('memcache_compress', false) == false ? 0 : MEMCACHE_COMPRESSED;
 
 		// Create the memcache connection
-		self::$_db = new Memcache;
-		self::$_db->addServer($server['host'], $server['port'], $this->_persistent);
+		static::$_db = new Memcache;
+		static::$_db->addserver($config->get('memcache_server_host', 'localhost'), $config->get('memcache_server_port', 11211), $this->_persistent);
 
-		$memcachetest = @self::$_db->connect($server['host'], $server['port']);
+		$memcachetest = @static::$_db->connect($server['host'], $server['port']);
```
Comment (csthomas) [inline libraries/joomla/cache/storage/memcache.php]: This line is incorrect!
$server is not defined, $memcachetest return false, and below throw exception.

This is reason why travis on PR #11576 failed.
https://travis-ci.org/joomla/joomla-cms/jobs/151946004

Comment (mbabker) [inline libraries/joomla/cache/storage/memcache.php]: Oh.  Oops.

Comment (rdeutz) [inline libraries/joomla/cache/storage/memcache.php]: Is this fixed somewhere or do we have to fix it, just curious.

**Comment by csthomas** [comment]
It is fixed at #11565.

**Comment by rdeutz** [comment]
@csthomas Thanks!

**Pull request merged**